### PR TITLE
Update minimum Python version from 3.5 to 3.8

### PR DESCRIPTION
**Pull request by benhoyt**
* Update references to the minimum Python version from 3.5 to 3.8
* Update setup.py `python_requires` version to 3.8
* Get rid of `framework._Metaclass` hack and use `__set_name__` instead
* Use `subprocess.run()`'s encoding argument now that we can
* Remove `pebble._json_loads()` now that `json.loads()` handles bytes
* `Pathlib.path.unlink()`'s "missing_ok" keyword argument
* Remove need for `unittest.skipIf` for Python < 3.7
* Couple of test tweaks now that dicts are ordered (as of Python 3.6)

Also added `test_run_error` from #868 (which is superceded by this PR).

**NOTE: This is a first pass, "minimum viable product" only.** In a subsequent release (2.1) I plan to do a full pass, updating everything to f-strings, variable type annotations, and so on.

### CODE REVIEW

File: ops/model.py [hunk 1]
```diff
@@ -2538,9 +2538,10 @@ def __init__(self, unit_name: Optional[str] = None,
         self._hook_is_running = ''
 
     def _run(self, *args: str, return_output: bool = False,
-             use_json: bool = False, input_stream: Optional[bytes] = None
+             use_json: bool = False, input_stream: Optional[str] = None
```
[the remark this answers is not in the document]
Comment (benhoyt) [reply ops/model.py:2551]: Yeah, it's an internal function and I updated the one caller to str, so I think it's fine to just accept str (it simplified the caller).

File: test/test_model.py [hunk 2]
```diff
@@ -301,14 +300,12 @@ def test_relation_data_modify_remote(self):
...
             data_repr = repr(rel_db1.data)
-        # the CountEqual and weird (and brittle) splitting is to accommodate python 3.5
-        # TODO: switch to assertEqual when we drop 3.5
-        self.assertCountEqual(
-            data_repr[1:-1].split(', '),
-            ["<ops.model.Unit myapp/0>: {}",
-             "<ops.model.Application myapp>: <n/a>",
-             "<ops.model.Unit remoteapp1/0>: {'host': 'remoteapp1/0'}",
-             "<ops.model.Application remoteapp1>: {'secret': 'cafedeadbeef'}"])
+        self.assertEqual(
```
Comment (rbarry82) [inline test/test_model.py:303]: Despite the comment, I'm not sure if this seems any less brittle (as a whitespace-sensitive multi-line string), but at least it's more legible.

**Review by rbarry82 — APPROVED** [review]
LGTM. Clean, simple cleanup.

Is this a first-pass only? There's probably a huge number of things which could be cleaned up for legibility with f-strings (especially some of the long `__repr__` and `log.error` messages), and probably some with `:=`, as much as I personally don't love it in Python all that much.

**Comment by benhoyt** [comment]
Thanks @rbarry82! Yes, this is definitely a first pass, minimum change to update various metadata (and a metaclass!) to Python 3.8. I plan to do the full pass with f-strings and additional good stuff (probably based on your previous PR) in a subsequent PR and release (likely 2.1).

I'll address your review comments shortly.

File: ops/storage.py [hunk 3]
```diff
@@ -372,7 +372,7 @@ def get(self, key: str) -> Any:
             CalledProcessError: if 'state-get' returns an error code.
         """
         # We don't capture stderr here so it can end up in debug logs.
-        p = _run(["state-get", key], stdout=subprocess.PIPE, check=True, universal_newlines=True)
+        p = _run(["state-get", key], stdout=subprocess.PIPE, check=True, encoding='utf-8')
```
[the remark this answers is not in the document]
Comment (benhoyt) [reply ops/storage.py]: Good call -- done.

File: test/test_model.py (hunk 2 above)
Comment (benhoyt) [reply test/test_model.py:303]: Yeah, I know! Six of one, half a dozen of the other... but I updated it anyway.

**Pull request merged**